### DISCUSSION TITLE
Can someone help me modify Ultimate SD Upscale to use the image seed used by the source file? (Python Question)

### DISCUSSION THREAD

**Discussion by marcsyp** [Mikubill/sd-webui-controlnet#1057] [Q&A] [answered]
Hi all --

I know this is not relevant to controlnet, specifically, but part of my controlnet workflow involves Ultimate SD Upscale and this is the most active developer community so I thought I would throw this out there.  I posted a [feature request on their repo](https://github.com/Coyote-A/ultimate-upscale-for-automatic1111/issues/73) but I don't have much hope that it will be addressed given how quiet that repo has been.  So I thought I could try to modify the USDU script to do what I want in the meantime, it would save me HOURS and HOURS of manual work.  

Only problem is I'm a C# developer and am very shaky in Python, and I don't know the SD ecosystem really at all.

TL;DR: 

Ultimate SD Upscaler doesn't work very well in batch mode for images that were generated using Dynamic Keywords or for multiple images from the same prompt coming from different seeds.  I would like to be able to use the exact same seed that was used for the generation of each image and have it also pull the correct prompt using the Dynamic Prompts from the webui.   I _think_ I can accomplish this by modifying the seed that is passed to Processed() in the following code:

          def run(self, p, _, tile_width, tile_height, mask_blur, padding, seams_fix_width, seams_fix_denoise, seams_fix_padding,
                      upscaler_index, save_upscaled_image, redraw_mode, save_seams_fix_image, seams_fix_mask_blur,
                      seams_fix_type, target_size_type, custom_width, custom_height, custom_scale):
          
                  # Init
                  processing.fix_seed(p)
                  devices.torch_gc()
          
                  p.do_not_save_grid = True
                  p.do_not_save_samples = True
                  p.inpaint_full_res = False
          
                  p.inpainting_fill = 1
          
                  seed = p.seed
          
                  # Init image
                  init_img = p.init_images[0]
                  if init_img == None:
                      return Processed(p, [], seed, "Empty image")
                  init_img = images.flatten(init_img, opts.img2img_background_color)
          
                  #override size
                  if target_size_type == 1:
                      p.width = custom_width
                      p.height = custom_height
                  if target_size_type == 2:
                      p.width = math.ceil((init_img.width * custom_scale) / 64) * 64
                      p.height = math.ceil((init_img.height * custom_scale) / 64) * 64
          
                  # Upscaling
                  upscaler = USDUpscaler(p, init_img, upscaler_index, save_upscaled_image, save_seams_fix_image, tile_width, tile_height)
                  upscaler.upscale()
                  
                  # Drawing
                  upscaler.setup_redraw(redraw_mode, padding, mask_blur)
                  upscaler.setup_seams_fix(seams_fix_padding, seams_fix_denoise, seams_fix_mask_blur, seams_fix_width, seams_fix_type)
                  upscaler.print_info()
                  upscaler.add_extra_info()
                  upscaler.process()
                  result_images = upscaler.result_images
          
                  return Processed(p, result_images, seed, upscaler.initial_info if upscaler.initial_info is not None else "")

Instead of "seed = p.seed" I would like to pull the seed from the image that is being upscaled.  It's totally fine if this is hardcoded into the script as a copy for now (no UI needed).  Am I on the right track?  Any hints?

**Comment by marcsyp** [comment, marked as answer]
Good news!

I poked around the extension code and the stable diffusion source code and managed to put together a few ideas that I ran past ChatGPT. We collaborated and I have successfully implemented all of the above in a custom forked version of the SDSU script. My comments on the solution are in the extension repo Discussion linked above.

I am currently running a batch USDU on 200 images with different prompts and seeds, and it is performing beautifully!

Cheers,
Marc
